pkg/operator/controllers/cpms/cpms_controller.go

```diff
@@ -0,0 +1,100 @@
+package cpms
+
+import (
+	"context"
+
+	machinev1 "github.com/openshift/api/machine/v1"
+	"github.com/sirupsen/logrus"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/Azure/ARO-RP/pkg/operator"
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	"github.com/Azure/ARO-RP/pkg/operator/controllers/base"
+)
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+const (
+	ControllerName         = "CPMSDeactivatorController"
+	SingletonCPMSName      = "cluster"
+	SingletonCPMSNamespace = "openshift-machine-api"
+)
+
+type Reconciler struct {
+	base.AROController
+}
+
+func NewReconciler(log *logrus.Entry, client client.Client) *Reconciler {
+	return &Reconciler{
+		AROController: base.AROController{
+			Log:    log,
+			Client: client,
+			Name:   ControllerName,
+		},
+	}
+}
+
+// CPMS reconciler will disable the cluster CPMS if `aro.cpms.enabled` is false or missing.
+func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	instance, err := r.GetCluster(ctx)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if instance.Spec.OperatorFlags.GetSimpleBoolean(operator.CPMSEnabled) {
+		r.Log.Infof("Flag %s is true, will not deactivate CPMS", operator.CPMSEnabled)
+		return reconcile.Result{}, nil
+	}
+
+	r.Log.Debug("running")
+	cpms := &machinev1.ControlPlaneMachineSet{}
+	err = r.Client.Get(
+		ctx,
+		types.NamespacedName{Name: SingletonCPMSName, Namespace: SingletonCPMSNamespace},
+		cpms,
+	)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			r.Log.Info("CPMS is not present on cluster, nothing to do")
+			return ctrl.Result{}, nil
+		}
+		r.Log.Errorf("Error when retrieving CPMS: %v", err)
+		return ctrl.Result{}, err
+	}
+
+	if cpms.Spec.State == machinev1.ControlPlaneMachineSetStateInactive {
+		r.Log.Info("CPMS is inactive, nothing to do")
+		return ctrl.Result{}, nil
+	}
+
+	// disable CPMS by deleting it
+	// https://docs.openshift.com/container-platform/4.12/machine_management/control_plane_machine_management/cpmso-disabling.html
+	return ctrl.Result{}, r.Client.Delete(ctx, cpms)
```

Comment:
Should we validate cpms is in an inactive state after deleting it?
I see we do that check beforehand, the next time the controller reconciles it will do the check.  
I'm just concerned about possible scenarios where cpms is getting recreated by something/someone, and our controller will just keep deleting it without raising a red flag.

Idea, maybe we can keep a tally somewhere of how many times this controller has deleted cpms, and if it's more than X(1?) generate an error.

Per the docs:
> Verification
Check the control plane machine set custom resource state. A result of Inactive indicates that the removal and replacement process is successful. A ControlPlaneMachineSet CR exists but is not activated.

Reply:
In the interest of keeping this controller in-line with controller recommendations (do one thing per reconcile loop), and the fact that this would introduce a dependency on this controller waiting for CPMSO to roll out a new Inactive CPMS, I'd rather not introduce a "wait" into the controller itself. 

The controller will reconcile against the newly-created inactive CPMS, and log that. IMO if we want to gather any data/metrics around CPMS being active/inactive, we should do so in a monitor rather than in this operator controller.